bootstrap/pkg/client/gcp/gcp.go

```diff
@@ -0,0 +1,632 @@
+}
+
+func (gcp *Gcp) Apply(resources kftypes.ResourceEnum, options map[string]interface{}) error {
+	updateDMErr := gcp.updateDM(resources, options)
+	if updateDMErr != nil {
+		return fmt.Errorf("gcp apply could not update deployment manager Error %v", updateDMErr)
+	}
+	secretsErr := gcp.createSecrets()
+	if secretsErr != nil {
+		return fmt.Errorf("gcp apply could not create secrets Error %v", secretsErr)
+	}
+	ks := gcp.Children[kftypes.KSONNET]
+	if ks != nil {
+		ksApplyErr := ks.Apply(resources, options)
+		if ksApplyErr != nil {
+			return fmt.Errorf("gcp apply failed for %v: %v", string(kftypes.KSONNET), ksApplyErr)
+		}
+	} else {
+		return fmt.Errorf("%v not in Children", string(kftypes.KSONNET))
+	}
+	return nil
+}
+
+func (gcp *Gcp) Delete(resources kftypes.ResourceEnum, options map[string]interface{}) error {
+	ks := gcp.Children[kftypes.KSONNET]
+	if ks != nil {
+		ksDeleteErr := ks.Delete(resources, options)
+		if ksDeleteErr != nil {
+			return fmt.Errorf("gcp delete failed for %v: %v", string(kftypes.KSONNET), ksDeleteErr)
+		}
+	} else {
+		return fmt.Errorf("%v not in Children", string(kftypes.KSONNET))
+	}
+	return nil
+}
+
+func (gcp *Gcp) copyFile(source string, dest string) error {
+	from, err := os.Open(source)
+	if err != nil {
+		return fmt.Errorf("cannot create directory %v", err)
+	}
+	defer from.Close()
+	to, err := os.OpenFile(dest, os.O_RDWR|os.O_CREATE, 0666)
+	if err != nil {
+		return fmt.Errorf("cannot create dest file %v  Error %v", dest, err)
+	}
+	defer to.Close()
+	_, err = io.Copy(to, from)
+	if err != nil {
+		return fmt.Errorf("copy failed source %v dest %v Error %v", source, dest, err)
+	}
+
+	return nil
+}
+
+func (gcp *Gcp) generateKsonnet(options map[string]interface{}) error {
+	email := gcp.GcpApp.Spec.Email
+	if options[string(kftypes.EMAIL)] != nil {
+		email = options[string(kftypes.EMAIL)].(string)
+		if email == "" {
+			return fmt.Errorf("email parameter required for cert-manager")
+		}
+	}
+	ipName := gcp.GcpApp.Spec.IpName
+	if options[string(kftypes.IPNAME)] != nil {
+		ipName = options[string(kftypes.IPNAME)].(string)
+		if ipName == "" {
+			return fmt.Errorf("ipName parameter required for iap-ingress")
+		}
+	}
+	hostname := gcp.GcpApp.Spec.Hostname
+	if options[string(kftypes.HOSTNAME)] != nil {
+		hostname = options[string(kftypes.HOSTNAME)].(string)
+		if hostname == "" {
+			return fmt.Errorf("hostname parameter required for iap-ingress")
+		}
+	}
+	project := gcp.GcpApp.Spec.Project
+	if options[string(kftypes.PROJECT)] != nil {
+		project = options[string(kftypes.PROJECT)].(string)
+		if project == "" {
+			return fmt.Errorf("project parameter required for iam_bindings")
+		}
+	}
+	zone := gcp.GcpApp.Spec.Zone
+	if options[string(kftypes.ZONE)] != nil {
+		zone = options[string(kftypes.ZONE)].(string)
+		if zone == "" {
+			return fmt.Errorf("zone parameter required for iam_bindings")
+		}
+	}
+	kstypes.DefaultPackages = append(kstypes.DefaultPackages, []string{"gcp"}...)
```

Comment:
Might be better to keep ```DefaultPackages DefaultComponents DefaultParameters``` unchanged.
We can make a copy and modify on new instance?

Reply:
Added 2 compatible config entries:
https://github.com/kubeflow/kubeflow/blob/master/bootstrap/config/kfctl_iap.yaml
https://github.com/kubeflow/kubeflow/blob/master/bootstrap/config/kfctl_basic_auth.yaml

Can we load ksonnet config from these 2 files instead of build with go logics?
So we can publish other custom entries without rebuild go binary.